core/capabilities/remote/target/request/caller_request.go

```diff
@@ -0,0 +1,160 @@
+		return nil, errors.New("remote capability info missing DON")
+	}
+
+	rawRequest, err := pb.MarshalCapabilityRequest(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal capability request: %w", err)
+	}
+
+	tc, err := transmission.ExtractTransmissionConfig(req.Config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract transmission config from request config: %w", err)
+	}
+
+	peerIDToTransmissionDelay, err := transmission.GetPeerIDToTransmissionDelay(remoteCapabilityDonInfo.Members, localDonInfo.Config.SharedSecret,
+		messageID, tc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get peer ID to transmission delay: %w", err)
+	}
+
+	responseReceived := make(map[p2ptypes.PeerID]bool)
+	for peerID, delay := range peerIDToTransmissionDelay {
+		responseReceived[peerID] = false
+		go func(peerID ragep2ptypes.PeerID, delay time.Duration) {
+			message := &types.MessageBody{
+				CapabilityId:    remoteCapabilityInfo.ID,
+				CapabilityDonId: remoteCapabilityDonInfo.ID,
+				CallerDonId:     localDonInfo.ID,
+				Method:          types.MethodExecute,
+				Payload:         rawRequest,
+				MessageId:       []byte(messageID),
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(delay):
+				err := dispatcher.Send(peerID, message)
+				if err != nil {
+					lggr.Errorw("failed to send message", "peerID", peerID, "err", err)
+				}
+			}
+		}(peerID, delay)
+	}
+
+	return &callerRequest{
+		createdAt:                  time.Now(),
+		requestTimeout:             requestTimeout,
+		requiredIdenticalResponses: int(remoteCapabilityDonInfo.F + 1),
+		responseIDCount:            make(map[[32]byte]int),
+		errorCount:                 make(map[string]int),
+		responseReceived:           responseReceived,
+		responseCh:                 make(chan commoncap.CapabilityResponse, 1),
+	}, nil
+}
+
+func (c *callerRequest) ResponseChan() <-chan commoncap.CapabilityResponse {
+	return c.responseCh
+}
+
+func (c *callerRequest) Expired() bool {
+	return time.Since(c.createdAt) > c.requestTimeout
+}
+
+func (c *callerRequest) Cancel(reason string) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	if !c.respSent {
+		c.sendResponse(commoncap.CapabilityResponse{Err: errors.New(reason)})
+	}
+}
+
+// TODO OnMessage assumes that only one response is received from each peer, if streaming responses need to be supported this will need to be updated
+func (c *callerRequest) OnMessage(_ context.Context, msg *types.MessageBody) error {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	if msg.Sender == nil {
+		return fmt.Errorf("sender missing from message")
+	}
+
+	sender := remote.ToPeerID(msg.Sender)
+
+	if _, ok := c.responseReceived[sender]; !ok {
+		return fmt.Errorf("response from peer %s not expected", sender)
+	}
+
+	if c.responseReceived[sender] {
+		return fmt.Errorf("response from peer %s already received", sender)
+	}
+
+	c.responseReceived[sender] = true
+
+	if msg.Error == types.Error_OK {
```

Comment:
this is odd. what is an Error_OK? does that mean no error? it seems like an oxymoron.

Reply:
I'll defer to @bolekk on that one, guess it could be StatusCode.

Reply:
yeah, it should be called status code, I can rename in a separate PR